numpy/core/src/multiarray/einsum.c.src

```diff
@@ -536,6 +536,10 @@ unbuffered_loop_nop1_ndim2(NpyIter *iter)
     for (coord = shape[1]; coord > 0; --coord) {
         sop(1, ptrs[0], strides[0], shape[0]);
 
+        if(PyErr_Occurred()){
+            return -1;
+        }
```

Comment:
I don't think calling `PyErr_Occurred` is safe from within `NPY_BEGIN_THREADS_THRESHOLDED`, but hopefully someone who remembers more about the threading (AKA releasing the GIL) can review to confirm or refute.

I suspect the real error here is that `NPY_BEGIN_THREADS_THRESHOLDED` is not safe to call at all when handling object arrays.

Reply:
I didn't consider this. 
Perhaps calling `sop` can be split into two functions based on `type_num`. one function releases the GIL were the other does not, and checks for `PyError_Occured`.

Reply:
See also gh-18450 for how the error should be checked here.

Reply:
I think the problem of API calls inside the `NPY_BEGIN_THREADS` macros  is resolved: it will only release the GIL and check error conditions if the `needs_api` is False.